sound/soc/sof/topology.c

```diff
@@ -2464,15 +2465,180 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+	struct snd_soc_dai_link *dai_link;
 	const struct firmware *fw;
+	bool load_default_tplg = false;
+	unsigned long tplg_mask = 0;
+	char platform[4];
+	char *tplg_name;
+	int tplg_num = 0;
+	int tplg_dev;
 	int ret;
+	int i;
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
-	ret = request_firmware(&fw, file, scomp->dev);
+	tplg_name = devm_kasprintf(sdev->dev, GFP_KERNEL,
+				   "%s", file);
+	if (sdev->pdata->ipc_type == SOF_IPC_TYPE_3)
+		goto legacy_tplg;
+
+	ret = sscanf(sof_pdata->tplg_filename, "sof-%3s-*.tplg", platform);
+	if (ret != 1 || !is_platform_support_separated_tplg(platform))
+		goto legacy_tplg;
+
+	for_each_card_prelinks(scomp->card, i, dai_link) {
+		char *tplg_device;
+
+		if (tplg_num >= MAX_TPLG_NUM) {
+			dev_err(scomp->dev,
+				"Invalid tplg_num %d, check what happened\n", tplg_num);
+			return -EINVAL;
+		}
+
+		dev_dbg(scomp->dev, "dai_link %s id %d\n", dai_link->name, dai_link->id);
+		if (strstr(dai_link->name, "SimpleJack")) {
+			tplg_dev = TPLG_DEVICE_SDW_JACK;
+			tplg_device = "sdca-jack";
+		} else if (strstr(dai_link->name, "SmartAmp")) {
+			tplg_dev = TPLG_DEVICE_SDW_AMP;
+			tplg_device = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						     "sdca-%damp", dai_link->num_cpus);
+			if (!tplg_device)
+				return -ENOMEM;
+		} else if (strstr(dai_link->name, "SmartMic")) {
+			tplg_dev = TPLG_DEVICE_SDW_MIC;
+			tplg_device = "sdca-mic";
+		} else if (strstr(dai_link->name, "dmic")) {
+			if (strstr(file, "-2ch")) {
+				tplg_device = "dmic-2ch";
+				tplg_name = strremove(tplg_name, "-2ch");
+			} else if (strstr(file, "-4ch")) {
+				tplg_device = "dmic-4ch";
+				tplg_name = strremove(tplg_name, "-4ch");
+			} else {
+				dev_warn(scomp->dev,
+					 "only -2ch and -4ch are supported for dmic\n");
+				continue;
+			}
+			tplg_dev = TPLG_DEVICE_HOST_DMIC;
+		} else if (strstr(dai_link->name, "iDisp")) {
+			tplg_dev = TPLG_DEVICE_HDMI;
+			tplg_device = "sdca-hdmi";
+
+		} else {
+			/* The dai link is not supported by sperated tplg yet */
+			load_default_tplg = true;
+			continue;
+		}
+		if (tplg_mask & BIT(tplg_dev))
+			continue;
+		tplg_mask |= BIT(tplg_dev);
+		tplg_files[tplg_num].be_id = dai_link->id;
+		tplg_files[tplg_num].device = tplg_device;
+		tplg_num++;
+	}
+	dev_dbg(scomp->dev, "tplg_mask %#lx tplg_num %d\n", tplg_mask, tplg_num);
+
+	/* Currently, only SDCA topology supported */
+	if (!(tplg_mask & SDCA_DEVICE_MASK)) {
+		/* Restore the default firmware name */
+		tplg_name = (char *)file;
+		goto legacy_tplg;
+	}
+
+	for (i = 0; i < tplg_num; i++) {
+		tplg_files[i].file = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						    "%s/sof-%s-%s-id%d.tplg",
+						    sof_pdata->tplg_filename_prefix, platform,
+						    tplg_files[i].device,
+						    tplg_files[i].be_id);
+		if (!tplg_files[i].file)
+			return -ENOMEM;
+
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i].file);
+		ret = request_firmware(&fw, tplg_files[i].file, scomp->dev);
```

Comment:
and the original name is altered, the -2/4ch is dropped, so we are not going to load the right tplg in some cases.

Reply:
> firmware_request_nowarn() to handle the fallback in case the fragmented tplg is not present, but what will happen if you load two fragment thn the third is missing and try to load the monolithic over it?
> 

It will return error if the second or later topology is not present. My idea is that if we can't find the first fragmented topology, it means we don't have the required version of topologies, then we need to fall back to use the original topology. However, if we find the first fragmented topology, it means we use the new version of topology, and there should be something wrong if we don't find any required fragmented topology.

> and the original name is altered, the -2/4ch is dropped, so we are not going to load the right tplg in some cases.

I will restore the file name in this case.